Source/Tests/LinuxSyscalls/Syscalls.cpp

```diff
@@ -226,89 +211,156 @@ static bool IsSupportedByInterpreter(std::string const &Filename) {
+
+  // We need to first determine the file size through fstat.
+  struct stat buf{};
+  if (fstat(FD, &buf) == -1) {
+    // Couldn't get size.
+    return false;
+  }
+
+  // Is the file large enough for shebang
+  if (buf.st_size <= 2)
+    return false;
+
+  // The maximum length of the shebang line is `#!` + 255 chars
+  std::vector<char> Header(std::min(257l, buf.st_size));
+
+  // Read the header so we can tell if it is a supported ELF file.
+  // Can't adjust file offset, so use pread.
+  if (pread(FD, &Header.at(0), Header.size(), 0) != Header.size()) {
+    // Couldn't read
+    LogMan::Msg::EFmt("Couldn't read potential shebang FD");
+    return false;
+  }
+
+  return IsShebangFile(&Header.at(0), Header.size());
+}
+
+uint64_t ExecveHandler(const char *pathname, char* const* argv, char* const* envp, ExecveAtArgs &Args) {
   std::string Filename{};
 
   std::error_code ec;
   std::string RootFS = FEX::HLE::_SyscallHandler->RootFSPath();
-
-  // Check the rootfs if it is available first
-  if (pathname[0] == '/') {
-    auto Path = FEX::HLE::_SyscallHandler->FM.GetEmulatedPath(pathname, true);
-    if (!Path.empty() && std::filesystem::exists(Path, ec)) {
-      Filename = Path;
+  ELFLoader::ELFContainer::ELFType Type{};
+  bool IsFDExec{};
+  std::string FDExecEnv;
+
+  if ((Args.flags & AT_EMPTY_PATH) && strlen(pathname) == 0) {
+    // AT_EMPTY_PATH is only used if the pathname is empty.
+    Type = ELFLoader::ELFContainer::GetELFType(Args.dirfd);
+
+    if (!IsShebangFD(Args.dirfd) && Type == ELFLoader::ELFContainer::ELFType::TYPE_NONE) {
+      // If our interpeter doesn't support this file format AND ELF format is NONE then ENOEXEC
+      // binfmt_misc could end up handling this case but we can't know that without parsing binfmt_misc ourselves
+      // Return -ENOEXEC until proven otherwise
+      return -ENOEXEC;
+    }
+    IsFDExec = true;
+  }
+  else
+  {
+    // Check the rootfs if it is available first
+    if (pathname[0] == '/') {
+      auto Path = FEX::HLE::_SyscallHandler->FM.GetEmulatedPath(pathname, true);
+      if (!Path.empty() && std::filesystem::exists(Path, ec)) {
+        Filename = Path;
+      }
+      else {
+        Filename = pathname;
+      }
     }
     else {
       Filename = pathname;
     }
-  }
-  else {
-    Filename = pathname;
-  }
 
-  bool exists = std::filesystem::exists(Filename, ec);
-  if (ec || !exists) {
-    return -ENOENT;
-  }
+    bool exists = std::filesystem::exists(Filename, ec);
+    if (ec || !exists) {
+      return -ENOENT;
+    }
+
+    int pid = getpid();
 
-  int pid = getpid();
+    char PidSelfPath[50];
+    snprintf(PidSelfPath, 50, "/proc/%i/exe", pid);
 
-  char PidSelfPath[50];
-  snprintf(PidSelfPath, 50, "/proc/%i/exe", pid);
+    if (strcmp(pathname, "/proc/self/exe") == 0 ||
+        strcmp(pathname, "/proc/thread-self/exe") == 0 ||
+        strcmp(pathname, PidSelfPath) == 0) {
+      // If pointing to self then redirect to the application
```

Comment:
What exactly is the scenario described here? Is it "FEX runs `java`, and `java` then calls `execve` on a symlink to its own executable to start the guest Java application"?

At that point, what is `/proc/self/exe` actually pointing to - FEX or `java`? Comment needs to be a bit clearer, I think.

Reply:
Updated comment.